### ISSUE TITLE
[HOLD for payment 2023-02-03] [$4000] Chat - The Paragraphs are not displayed and some spaces missing between text after copied test from the opened HTML file

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#12155]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Issue found when executing PR** https://github.com/Expensify/App/pull/11926

## Action Performed:
1. Go to staging.new.expensify.com
2. Log in with any account
3. Select any chat
4. Open HTML file in another tab
5. Copy and paste as is
6. Send it

## Expected Result:
The Paragraphs and spaces between text are displayed after  copied test from the opened HTML file

## Actual Result:
The Paragraphs are not displayed and some spaces missing between text after copied test from the opened HTML file

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.2.19.2
 
**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/198123365-fb5fdc99-134c-4e5a-ad93-7b051121093e.mp4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Applause - Internal Team

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ca19ac6a046e22d6</li>
        <li>Upwork Job ID: 1613324508306006016</li>
        <li>Last Price Increase: 2023-01-25</li>
    </ul>
</details>

**Comment by Gonals** [comment]
Marking as external!

**Comment by Santhosh-Sellavel** [comment]
We are still waiting for proposals here! ~Maybe we should double this~  @davidcardoza please post the Upwork job! Maybe we could get some proposals! I think we could start this job for $500 or $1000!

cc: @tgolen!

**Comment by tgolen** [comment]
Still waiting for proposals.

**Comment by puneetlath** [comment]
@tgolen To help us clear out the large backlog of /App bugs, we're putting the spotlight every bug in the repo already than 4 weeks old. To help unblock the roadmap and get our bug pipeline back in equilibrium, can you:  
 
- Decide whether any proposals currently meet our guidelines and can be approved as-is 
- For any that can't, please take this issue internal and treat it as one of your highest priorities 
- If you have any questions, don't hesitate to start a discussion in #bug-zero

**Comment by fabiocberg** [comment]
Hi, my name is Fabio, I've been working as a web, backend and mobile developer for over 16 years.
I can help you with your paste issue.
The problem is with your input component. This is accepting only single lines, so after pasting the text, it automatically removes break lines.
It can be fixed by creating another input component (or adapting the current one) to accept multiline text.

Best regards,

Fabio Bergmann

**Comment by tgolen** [comment]
@fabiocberg Thank you for your proposal. Can you please give more details about the root of the problem (add links to the code where it shows it's only accepting single lines) and then also show the specific code changes you would make to solve the problem?

**Comment by tjferriss** [comment]
Reviewing the issue based on the [weekly bug update chore](https://stackoverflow.com/c/expensify/questions/8208):

@tgolen this is one of the oldest issues in the /App repo. To help us clear out the large backlog of bugs, can you:
- Decide whether any proposals currently meet our guidelines and can be approved as-is
- For any that can't, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #bug-zero

**Comment by cead22** [comment]
- Adding `multiline={true}` to the RNTextInput [here](https://github.com/Expensify/App/blob/00d7c619c/src/components/Composer/index.js#L390) didn't work
- Adding `multiline={true}` to TextInput [here](https://github.com/Expensify/App/blob/00d7c619c/src/components/RNTextInput.js#L27) didn't work either

**Comment by Santhosh-Sellavel** [comment]
Looking for proposals

**Comment by eh2077** [comment]
Proposal

We can support pasting HTML `p` tag and unordered list in `ExpensiMark.js` of [expensify-common](https://github.com/Expensify/expensify-common). Two changes will be added into `ExpensiMark.js`
1. Add space for unordered list item https://github.com/Expensify/expensify-common/blob/db2c245c1e6b39232092b1415b9235fadda2a529/lib/ExpensiMark.js#L226-L231
```diff
                name: 'heading1',
                regex: /\s*<(h1)(?:"[^"]*"|'[^']*'|[^'">])*>(.*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))\s*/gi,
                replacement: '\n# $2\n',
            },
+            {
+                name: 'listItem',
+                regex: /\s*<(li)(?:"[^"]*"|'[^']*'|[^'">])*>(.*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))\s*/gi,
+                replacement: '<li>  $2</li>',
+            },
             {
                 name: 'italic',
```
2. Add line break for HTML `p` and `li` tag https://github.com/Expensify/expensify-common/blob/db2c245c1e6b39232092b1415b9235fadda2a529/lib/ExpensiMark.js#L465-L474
```diff
        this.htmlToMarkdownRules.forEach((rule) => {
            // Pre-processes input HTML before applying regex
            if (rule.pre) {
                generatedMarkdown = rule.pre(generatedMarkdown);
            }
            generatedMarkdown = generatedMarkdown.replace(rule.regex, rule.replacement);
        });
-        generatedMarkdown = generatedMarkdown.replace(/<div.*?>|<\/div>|<comment.*?>|\n<\/comment>|<\/comment>|<h2>|<\/h2>|<h3>|<\/h3>|<h4>|<\/h4>|<h5>|<\/h5>|<h6>|<\/h6>/gm, '[block]');
+        generatedMarkdown = generatedMarkdown.replace(/<div.*?>|<\/div>|<comment.*?>|\n<\/comment>|<\/comment>|<h2>|<\/h2>|<h3>|<\/h3>|<h4>|<\/h4>|<h5>|<\/h5>|<h6>|<\/h6>|<p>|<\/p>|<li>|<\/li>/gm, '[block]');
         return Str.htmlDecode(this.replaceBlockWithNewLine(Str.stripHTML(generatedMarkdown)));
```
Video demo about the change

https://user-images.githubusercontent.com/117511920/204067360-c1d011b6-669c-4f64-8ae6-ac2d1d755f7d.mov

Test html
```html
<!DOCTYPE html>
<html>
<body>
    <h1>HTML paragraphs and unordered list copy test</h1>
    <p>This is a HTML paragraph</p>
    <br>
    <p>An Unordered HTML List</p>
    <ul>
        <li><a href="https://example.com">Coffee</a> -- Coffee</li>
        <li><a href="https://example.com">Tea</a> -- Tea</li>
        <li><a href="https://example.com">Milk</a> -- Milk</li>
    </ul>
</body>
</html>

```

To have least changes in `ExpensiMark.js` and keep it simple, this proposal converts HTML list items to new lines with 2 prefix spaces instead of markdown list. 

I'll be happy to hear feedback from you! cc @Santhosh-Sellavel

**Comment by Santhosh-Sellavel** [comment]
We have a proposal and will review it shortly!

**Comment by Santhosh-Sellavel** [comment]
@tgolen this is more like a feature than a bug, shall we proceed with this?

@eh2077 Not working for me, can you check and update please thanks!

**Comment by tgolen** [comment]
I agree that this is a new feature. @davidcardoza I'm going to place this on HOLD until we get through What'sAppQuality 👍

**Comment by eh2077** [comment]
> @eh2077 Not working for me, can you check and update please thanks!

Hi @Santhosh-Sellavel, thanks for checking the proposal! 

I apply the diff of upstream lib `expensify-common` to App project through following steps.
First, clone `expensify-common` to local disk and apply following diff
```diff
diff --git a/lib/ExpensiMark.js b/lib/ExpensiMark.js
index 8eb6b92..f9ad8cf 100644
--- a/lib/ExpensiMark.js
+++ b/lib/ExpensiMark.js
@@ -227,6 +227,11 @@ export default class ExpensiMark {
                 regex: /\s*<(h1)(?:"[^"]*"|'[^']*'|[^'">])*>(.*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))\s*/gi,
                 replacement: '\n# $2\n',
             },
+            {
+                name: 'listItem',
+                regex: /\s*<(li)(?:"[^"]*"|'[^']*'|[^'">])*>(.*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))\s*/gi,
+                replacement: '<li>  $2</li>',
+            },
             {
                 name: 'italic',
                 regex: /<(em|i)(?:"[^"]*"|'[^']*'|[^'">])*>(.*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))/gi,
@@ -469,7 +474,7 @@ export default class ExpensiMark {
             }
             generatedMarkdown = generatedMarkdown.replace(rule.regex, rule.replacement);
         });
-        generatedMarkdown = generatedMarkdown.replace(/<div.*?>|<\/div>|<comment.*?>|\n<\/comment>|<\/comment>|<h2>|<\/h2>|<h3>|<\/h3>|<h4>|<\/h4>|<h5>|<\/h5>|<h6>|<\/h6>/gm, '[block]');
+        generatedMarkdown = generatedMarkdown.replace(/<div.*?>|<\/div>|<comment.*?>|\n<\/comment>|<\/comment>|<h2>|<\/h2>|<h3>|<\/h3>|<h4>|<\/h4>|<h5>|<\/h5>|<h6>|<\/h6>|<p>|<\/p>|<li>|<\/li>/gm, '[block]');
         return Str.htmlDecode(this.replaceBlockWithNewLine(Str.stripHTML(generatedMarkdown)));
     }
```
Second, switch to local dependency in the `package.json` file of App project
```diff
diff --git a/package.json b/package.json
index fb2e1d98cf..fe2e41901e 100644
--- a/package.json
+++ b/package.json
@@ -67,7 +67,7 @@
     "dom-serializer": "^0.2.2",
     "domhandler": "^4.3.0",
     "dotenv": "^8.2.0",
-    "expensify-common": "git+https://github.com/Expensify/expensify-common.git#87d6599f2fadd3601067ac9cfa5d10681fa34811",
+    "expensify-common": "file:/Users/erich/path/to/Expensify/expensify-common",
     "fbjs": "^3.0.2",
     "file-loader": "^6.0.0",
     "html-entities": "^1.3.1",
```
Last, run `npm install --save` and start the App.

Hope this helps though this new feature is currently marked as hold.

**Comment by Santhosh-Sellavel** [comment]
@eh2077 This is on hold, please focus on other [open](https://github.com/Expensify/App/issues?q=is%3Aissue+is%3Aopen+label%3A%22Help+Wanted%22) issues for now. When this is off-hold I will review this, thanks!

**Comment by tgolen** [comment]
I'm going to move this to a weekly issue while it's on HOLD

**Comment by tgolen** [comment]
Still on HOLD

**Comment by davidcardoza** [comment]
On HOLD until we get through WAQ

**Comment by tgolen** [comment]
Still on HOLD

**Comment by tgolen** [comment]
I'm taking this off HOLD now that WAQ is mostly behind us. I think that [this proposal](https://github.com/Expensify/App/issues/12155#issuecomment-1327955648) looks fine to me @eh2077 and @Santhosh-Sellavel

**Comment by Santhosh-Sellavel** [comment]
@tgolen Can you assign another C+ continue , I am unassigning myself here due to limited availability.

**Comment by tgolen** [comment]
OK, I reapplied the `external` label to get a new C+. @davidcardoza I think you can close [that duplicate upwork issue](https://github.com/Expensify/App/issues/12155#issuecomment-1379633837)

**Comment by thesahindia** [comment]
> I think that [this proposal](https://github.com/Expensify/App/issues/12155#issuecomment-1327955648) looks fine to me @eh2077 

Looks good to me as well!

**Comment by tgolen** [comment]
@davidcardoza can you please hire @eh2077 for this?

**Comment by tgolen** [comment]
bump @davidcardoza

**Comment by davidcardoza** [comment]
@eh2077 can you please apply for the job here https://www.upwork.com/jobs/~018589a0d15b72b606 ? Thanks!

**Comment by eh2077** [comment]
> @eh2077 can you please apply for the job here https://www.upwork.com/jobs/~018589a0d15b72b606 ? Thanks!

Hi @davidcardoza, I have applied the job. Thanks!

**Comment by davidcardoza** [comment]
Offer sent

**Comment by eh2077** [comment]
Hi @tgolen and @thesahindia , please help to review this PR https://github.com/Expensify/expensify-common/pull/490 when you're available. Thanks!

**Comment by tgolen** [comment]
Sure, I'm happy to review it. Next time, please just add me as a reviewer to the PR.

**Comment by eh2077** [comment]
### PR description draft

### Details

When replacing html with `<p>` and `<li>` tags, it misses newlines and some spaces. This PR fixes it by

- adding a newline when replacing `<p>` tag
- adding a newline and 2 leading spaces when replacing `<li>` tag


### Fixed Issues

$ https://github.com/Expensify/App/issues/12155   
PROPOSAL: https://github.com/Expensify/App/issues/12155#issuecomment-1327955648


### Tests


1. Login with any account and go to any chat room
2. Open/render HTML file in a browser 
```html
<!DOCTYPE html>
<html>
<body>
    <h1>HTML paragraphs and unordered list copy test</h1>
    <p>This is a HTML paragraph</p>
    <br>
    <p>An Unordered HTML List</p>
    <ul>
        <li><a href="https://example.com">Coffee</a> -- Coffee</li>
        <li><a href="https://example.com">Tea</a> -- Tea</li>
        <li><a href="https://example.com">Milk</a> -- Milk</li>
    </ul>
</body>
</html>
```
3. Copy and paste as is
4. Send it
5. Verify that `<p>` tag displays in newline and `<li>` item displays in newline with 2 prefix spaces

Note: For iOS native platform and Android native platform, only need to verify the sent text is displayed correctly because they don't support HTML pasting feature.

- [x] Verify that no errors appear in the JS console

### Offline tests


Same as tests, but offline

### QA Steps


1. Login with any account and go to any chat room
2. Open/render HTML file in a browser 
```html
<!DOCTYPE html>
<html>
<body>
    <h1>HTML paragraphs and unordered list copy test</h1>
    <p>This is a HTML paragraph</p>
    <br>
    <p>An Unordered HTML List</p>
    <ul>
        <li><a href="https://example.com">Coffee</a> -- Coffee</li>
        <li><a href="https://example.com">Tea</a> -- Tea</li>
        <li><a href="https://example.com">Milk</a> -- Milk</li>
    </ul>
</body>
</html>
```
3. Copy and paste as is
4. Send it
5. Verify that `<p>` tag displays in newline and `<li>` item displays in newline with 2 prefix spaces

Note: For iOS native platform and Android native platform, only need to verify the sent text is displayed correctly because they don't support HTML pasting feature.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist


- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/117511920/214559009-9c159b55-9139-4c2f-8273-284674bd8b7c.mov




</details>


<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/117511920/214559093-646b0cff-8ac8-49f4-8356-610652b3575a.mp4


</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/117511920/214559190-a317519f-0dda-4cb3-b82e-7567862dd350.mp4




</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/117511920/214559248-61b0c8b9-c47d-49b1-98b4-ab768617043e.mp4




</details>

<details>
<summary>iOS</summary>

Only need to verify the sent text is displayed correctly on iOS native platform because it doesn't support HTML pasting feature.

<img src="https://user-images.githubusercontent.com/117511920/214559309-4b17e4e5-538c-4a17-8d6a-9c64cc87ab3a.png" width="350">





</details>

<details>
<summary>Android</summary>

Only need to verify the sent text is displayed correctly on Android native platform because it doesn't support HTML pasting feature.

<img src="https://user-images.githubusercontent.com/117511920/214559359-9690bc37-20e7-47dd-bfcd-24d6186f0eba.png" width="350">



</details>

**Comment by eh2077** [comment]
Hi @davidcardoza , would you like to have a look on this when you have time?

**Comment by tgolen** [comment]
Reaching out to @davidcardoza to bump him directly

**Comment by davidcardoza** [comment]
@thesahindia Can you apply for the job here - https://www.upwork.com/jobs/~0159082963148d6a33  I was unable to find you in Upwork to hire you.

**Comment by tgolen** [comment]
bump @davidcardoza

**Comment by dylanexpensify** [comment]
@eh2077 can you apply please? 🙇

**Comment by dylanexpensify** [comment]
Alrighty, @eh2077 and @thesahindia payments sent and contracts ended! @davidcardoza I'll leave it to you to do the regression test bits!

**Comment by tgolen** [comment]
@davidcardoza this is still waiting on you

**Comment by tgolen** [comment]
Bumping @davidcardoza 1:1

**Comment by davidcardoza** [comment]
Regression test being added here - https://github.com/Expensify/Expensify/issues/275543

**Comment by tgolen** [comment]
Looks like this can be closed now
